core/transformer/src/test/standalone/IModelTransformerResumption.test.ts

```diff
@@ -0,0 +1,726 @@
+    if (this.owningTransformer === undefined)
+      throw Error("uninitialized, '_owningTransformer' must have been set before transformations");
+    ++this.owningTransformer.importedEntities;
+    return super.importElement(elementProps);
+  }
+}
+
+/** this class services two functions,
+ * 1. make sure transformations can be resumed by subclasses with different constructor argument types
+ * 2. count operations that should not be reperformed by a resumed transformation
+ */
+class CountingTransformer extends IModelTransformer {
+  public importedEntities = 0;
+  public exportedEntities = 0;
+  constructor(opts: {
+    source: IModelDb;
+    target: IModelDb;
+    options?: IModelTransformOptions;
+  }) {
+    super(
+      opts.source,
+      new CountingImporter(opts.target),
+      opts.options
+    );
+    (this.importer as CountingImporter).owningTransformer = this;
+  }
+  public override onExportElement(sourceElement: Element) {
+    ++this.exportedEntities;
+    return super.onExportElement(sourceElement);
+  }
+  public override onExportRelationship(sourceRelationship: Relationship) {
+    ++this.exportedEntities;
+    return super.onExportRelationship(sourceRelationship);
+  }
+}
+
+/** a transformer that executes a callback after X element exports */
+class CountdownTransformer extends IModelTransformer {
+  public relationshipExportsUntilCall: number | undefined;
+  public elementExportsUntilCall: number | undefined;
+  public callback: (() => Promise<void>) | undefined;
+  public constructor(...args: ConstructorParameters<typeof IModelTransformer>) {
+    super(...args);
+    const _this = this; // eslint-disable-line @typescript-eslint/no-this-alias
+    const oldExportElem = this.exporter.exportElement; // eslint-disable-line @typescript-eslint/unbound-method
+    this.exporter.exportElement = async function (...args) {
+      if (_this.elementExportsUntilCall === 0) await _this.callback?.();
+      if (_this.elementExportsUntilCall !== undefined)
+        _this.elementExportsUntilCall--;
+      return oldExportElem.call(this, ...args);
+    };
+    const oldExportRel = this.exporter.exportRelationship; // eslint-disable-line @typescript-eslint/unbound-method
+    this.exporter.exportRelationship = async function (...args) {
+      if (_this.relationshipExportsUntilCall === 0) await _this.callback?.();
+      if (_this.relationshipExportsUntilCall !== undefined)
+        _this.relationshipExportsUntilCall--;
+      return oldExportRel.call(this, ...args);
+    };
+  }
+}
+
+/** a transformer that crashes on the nth element export, set `elementExportsUntilCall` to control the count */
+class CountdownToCrashTransformer extends CountdownTransformer {
+  public constructor(...args: ConstructorParameters<typeof CountdownTransformer>) {
+    super(...args);
+    this.callback = () => { throw Error("crash"); };
+  }
+}
+
+/**
+ * Wraps all IModel native addon functions and constructors in a randomly throwing wrapper,
```

Comment:
Tests and randomness are generally not a good combo...

Reply:
the randomness is not used by default. It is behind a skipped/disabled test which you can activate to get some statistics of the speed of resumed transformations and whether contrived thrown errors throughout the stack generate a different result when resumed from a non-crashing transformation.

I enabled the test locally and ran several thousand transformations to test if resuming after various different crash points caused different results. I probably should put it somewhere, maybe in a work item, but I intend to run this in the transformer performance tests when those are setup. Just not as part of the unit test suite.

Reply:
Thanks for clarifying. Did it resume correctly when you tested locally?

Reply:
Yes. The test transforms without the crashes enabled, and with, checks that crashes did occur, and then confirms that iModels that were crashed and resumed as well as iModels that ran to completion without classes have the same content, using the  `assertIdentityTransformation` function which asserts equality of as much as it can about entities, relationships, properties, between two imodels using some remapping scheme such as that in the completed transformer